### DISCUSSION TITLE
Developments and Tests

### DISCUSSION THREAD

**Discussion by rdz-oss** [rdz-oss/BattyBirdNET-Analyzer#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

**Comment by spambake** [comment]
Recording is at 192kHz using a homebuild audio card and microphone amplifier. The mic position is poor - typical of many BirdNET pi installations.  On a condo deck with two sliding glass doors at right angles to each other. This distorts the sound somewhat at ultrasonic frequencies due to echos.

Results:
python3 analyze.py --min_conf 0.5 --overlap 0.0 --classifier ./checkpoints/bats/nabat-100-144kHz-200epochs.tflite --i ./testcalls2/ --o ./results2/
1	Spectrogram 1	1	1.0	2.0	150	15000	Corynorhinus townsendii_Townsend's big-eared bat	Townsend's big-eared bat	0.9907
2	Spectrogram 1	1	1.0	2.0	150	15000	Myotis austroriparius_Southeastern myotis	Southeastern myotis	0.9531
3	Spectrogram 1	1	3.0	4.0	150	15000	Myotis grisescens_Gray bat	Gray bat	0.8633
4	Spectrogram 1	1	4.0	5.0	150	15000	Myotis grisescens_Gray bat	Gray bat	0.9819
5	Spectrogram 1	1	5.0	6.0	150	15000	Myotis grisescens_Gray bat	Gray bat	0.7976

--overlap 0.1 - seems best on this file so far
1	Spectrogram 1	1	0.9	1.9	150	15000	Corynorhinus townsendii_Townsend's big-eared bat	Townsend's big-eared bat	0.9995
2	Spectrogram 1	1	3.6	4.6	150	15000	Myotis grisescens_Gray bat	Gray bat	0.9134
3	Spectrogram 1	1	4.5	5.5	150	15000	Myotis austroriparius_Southeastern myotis	Southeastern myotis	0.6156
4	Spectrogram 1	1	4.5	5.5	150	15000	Myotis grisescens_Gray bat	Gray bat	0.5756

--overlap 0.2
1	Spectrogram 1	1	0.8	1.8	150	15000	Corynorhinus townsendii_Townsend's big-eared bat	Townsend's big-eared bat	0.7928
2	Spectrogram 1	1	2.4000000000000004	3.4000000000000004	150	15000	Myotis austroriparius_Southeastern myotis	Southeastern myotis	0.5356
3	Spectrogram 1	1	3.2	4.2	150	15000	Myotis grisescens_Gray bat	Gray bat	0.9162
4	Spectrogram 1	1	3.2	4.2	150	15000	Myotis austroriparius_Southeastern myotis	Southeastern myotis	0.6803
5	Spectrogram 1	1	4.0	5.0	150	15000	Myotis grisescens_Gray bat	Gray bat	0.9819
6	Spectrogram 1	1	4.8	5.8	150	15000	Myotis grisescens_Gray bat	Gray bat	0.7470
7	Spectrogram 1	1	4.8	5.8	150	15000	Myotis austroriparius_Southeastern myotis	Southeastern myotis	0.5231

--overlap 0.3
1	Spectrogram 1	1	0.7	1.7	150	15000	Corynorhinus townsendii_Townsend's big-eared bat	Townsend's big-eared bat	0.8826
2	Spectrogram 1	1	2.8	3.8	150	15000	Myotis austroriparius_Southeastern myotis	Southeastern myotis	0.6939
3	Spectrogram 1	1	3.5	4.5	150	15000	Myotis austroriparius_Southeastern myotis	Southeastern myotis	0.8803
4	Spectrogram 1	1	3.5	4.5	150	15000	Myotis grisescens_Gray bat	Gray bat	0.8362
5	Spectrogram 1	1	4.2	5.2	150	15000	Myotis grisescens_Gray bat	Gray bat	0.9768
6	Spectrogram 1	1	4.2	5.2	150	15000	Myotis austroriparius_Southeastern myotis	Southeastern myotis	0.6600
7	Spectrogram 1	1	4.9	5.9	150	15000	Myotis grisescens_Gray bat	Gray bat	0.6675

I did not expect the 240 kHz to work since the recordings are at 192kHz. Let me know if this is incorrect. No results from this command so I didn't go further.  Changed config.py SAMPLE_RATE: int = 240000 and SIG_LENGTH: float = 0.6
python3 analyze.py --min_conf 0.5 --overlap 0.0 --classifier ./checkpoints/bats/nabat-100-240kHz-200epochs.tflite --i ./testcalls2/ --o ./results2/

**Comment by spambake** [comment]
I'm going to try altering a version of analyze.py to provide more readable output for multiple files that cover a full night.

**Reply by rdz-oss** [reply]
That is probably a good idea. I plan to write a new 'bat_identify.py' to have a smoother call interface and settings for bats. At this point I dont want to modify the analyze.py from BirdNET-Analyzer to a point where it would not be compatible anymore.

Also, It might make sense to offer classifiers for US-East and US-West separately. Perhaps even one for California in particular? That would automatically reduce some potential errors. Do you perhaps have a species list for your area?

**Reply by spambake** [reply]
I have an old list for my county but they've been doing lots of bat surveys for the last several year so I should be able to get an update.  I'll also put together lists for east and west North America as well as California. I wonder if I can just use --slist to point to one of these partial lists.

I'm not convinced the model is the cause of the problem I'm seeing, though it's possible.  Gray bat and SE myotis detections are occurring constantly, even in files with no bat calls and even with various sensitivity and overlap settings.  Possibilities are: model, frequency conversion artifacts, microphone location or the particular noise introduced by my sound card. I should be able to narrow it down in a few days.

**Reply by spambake** [reply]
I switched to a new machine and audio card, same microphone, and it's working much better. Many detections of Brazilian free-tailed bats with a few other local bats sprinkled in which fits the pattern for this time of year.  Occasional false detection as expected.  Kind of hard to verify detections.

I'll get my county list and a California list to you soon.  It's great you found those European bat calls.

**Reply by rdz-oss** [reply]
Thank you very much for assembling the lists! I will then make regional classifiers for the US !

The 'Bavarian'classifier  that I just uploaded into the repo  works well on my local data from Munich. I have a lot of myotis and pipistrelles, often flying in the same area near water. I compare it to the classifications made by the wildlife acoustics software on android. It is off sometimes, but my impression is it is a good basis to start moving to a monitoring system. Might actually work better than the wildlife acoustics classifier  (not validated).

 It is wrong about 1 in 20 times over the entire training set. So probably a bit more often in the real world with noise etc. The key will be to get more data, train the model cleverly with noise and to have a good heuristic for declaring a presence - such as a minimum of 5 consecutive seconds in which a the same species was detected. That would lose a few, but those that are registered will be very likely also present. It will be interesting to try this.

You might want to try the new bat_ident.py function. It should reduce the boiler plate code when calling the system. Put your files in the 'put-your-files-here' folder and use 

python3 bat_ident.py --area USA

and it will put the results of all files it finds in that folder into put-your-files-her/results . I started on a graphical version today but there was so much spaghetti code in there I got frustrated - for now.

Also - I included a new classifier for the US that was trained with up to 200 recordings per species. It is selected automatically if you use --area USA . There is also a classifier in the checkpoints/bats/v1.0 that was trained on the entire nabat data that you could use with the --classifier option

**Reply by spambake** [reply]
I'm really enjoying using this. Being able to identify bats with open source projects opens lots of possibilities. It would be great to eventually tie in with something like birdweather.com and track bat movements around the world.

Even with BirdNET-Pi I get plenty of false detections if I open to the full model with the species occurrence frequency setting. One of the problems with narrowing it down is that you can miss bats that didn't read the rules about where they're supposed to be.  Around here in my county we get a fair number of birds from the East US that get lost on their migration.  I wonder if bats do the same.

Tried the bat_ident.py and it worked well.  I noticed a difference with the new classifier but I'm not sure what that means yet.  This might work as well or better than any of the other software out there.

**Reply by rdz-oss** [reply]
It is fun to compare the results with what the app said when recording. The results tend to be more detailed - myotis myotis and pipistrellus pipistrellus fly at the same location and are segmented in a plausible way by the BattyBirdNET. It would be good to make a list of interesting features and prioritize soon.
Such as 
- getting the segment.py to work for bats - cutting the recordings into pieces according to detections. This could also be a useful feature for generating ( a lot) more training data
- a graphical user interface adaptation 
- Including more areas on the planet - ChiroVox would allow India/Malysia/China  and more. Not enough data for a great classifier but maybe enough for a useful one at this point
- Integration into BirdNET-Pi
- I also want to write up a paper  on training data set and method

Maybe there are people who would assemble the data in the right way for ChiroVox East Asia or work on  the  GUI

**Comment by spambake** [comment]
Here's a couple of bat lists.  Not sure what format you want.  Let me know if you want a different format. I can also do lists as four letter codes.

It'll take me a while to get the east - west divisions.  A little harder than I thought.  I'll probably make state lists and merge them for different regions.

Marin County, CA USA
From old list and NABAT data
Antrozous pallidus_Pallid bat
Corynorhinus townsendii_Townsend's big-eared bat
Eptesicus fuscus_Big brown bat
Eumops perotis_Greater mastiff bat
Lasionycteris noctivagans_Silver-haired bat
Lasiurus blossevillii_Western red bat
Lasiurus cinereus_Hoary ba
Myotis californicus_California bat
Myotis evotis_Western long-eared bat
Myotis lucifugus_Little brown bat
Myotis thysanodes_Fringed bat
Myotis volans_Long-legged bat
Myotis yumanensis_Yuma bat
Parastrellus hesperus_Canyon bat
Tadarida brasiliensis_Brazilian free-tailed bat

California list from NABAT data:
Antrozous pallidus_Pallid bat
Corynorhinus townsendii_Townsend's big-eared bat
Eptesicus fuscus_Big brown bat
Euderma maculatum_Spotted bat
Eumops perotis_Greater mastiff bat
Lasionycteris noctivagans_Silver-haired bat
Lasiurus blossevillii_Western red bat
Lasiurus cinereus_Hoary bat
Lasiurus xanthinus_Western yellow bat
Myotis californicus_California bat
Myotis ciliolabrum_Western small-footed myotis
Myotis evotis_Western long-eared bat
Myotis lucifugus_Little brown bat
Myotis thysanodes_Fringed bat
Myotis volans_Long-legged bat
Myotis yumanensis_Yuma bat
Nyctinomops femorosaccus_Pocketed free-tailed bat
Nyctinomops macrotis_Big free-tailed bat
Parastrellus hesperus_Canyon bat
Tadarida brasiliensis_Brazilian free-tailed bat

**Reply by rdz-oss** [reply]
Hi! Thanks again for the list. I added a classifier for MarinCounty.
python3 bat_ident.py --area MarinCounty
should work now with species on your list.

**Reply by rdz-oss** [reply]
You can get segmented .wav files by species with a spectrogram plot (if you install sox) by using
python3 bat_ident.py --area MarinCounty --spectrum on
I will provide classifiers for 288kHz also in maybe two weeks.

**Reply by spambake** [reply]
Thanks for the local classifier and the segments of .wav files with spectrograms.  It should make it much easier to verify detections. 
I'll be traveling for the next week without sufficient tech gear so I'll test it out when I get back.

**Reply by rdz-oss** [reply]
Hi @spambake ,

you want to let you known that there are now USA-EST and USA-WEST classifiers available. Also detection can now be limited from dusk to dawn. Robustness against noise (on audiomoth) also increased significantly. You can also choose to denoise tha audiomoth recordings in the settings.
Hope that is useful,

Best regards,
richard

**Reply by spambake** [reply]
That's very useful.  I'll do a fresh install and give it a try.  It's good to see some additional interest in the project.

**Comment by rdz-oss** [comment]
Once you ran bat_ident.py and have a result file, you can obtain the segmented audio files sorted by detected species using 
``` sh
python3 segments.py --audio path/to/input/audio --results path/to/input/audio/results --o path/to/input/audio/segments
```
If you have put your files in the folder 'put-your-files-here', the you can use
``` sh
python3 segments.py
```
If you use the above folder and know that you want to extract the segements, you can use
``` sh
python3 bat_ident.py --segment on
```
and it will generate the segemnts to 'put-your-files-here/segments'.

**Comment by rdz-oss** [comment]
Check out the online demo for small files ( a few seconds) now on huggingface.co:
https://huggingface.co/spaces/Amazetl/BattyBirdNET-Analyze-Demo
If you are from the UK, why not cross check to batdetect2
https://huggingface.co/spaces/macaodha/batdetect2
Enjoy and let me know how it works for you! 

I made a mistake when changing to 1444kHz version. In effect, it was running a hybrid between 144 and 256 kHz. Should be fixed now. My appologies. Now it should work .

**Reply by NoxTheXelor** [reply]
![image](https://github.com/rdz-oss/BattyBirdNET-Analyzer/assets/26748640/39c5f8f4-21be-4a64-9567-913ba65818a8)
It seems there is an error on the HuggingFace website

**Reply by rdz-oss** [reply]
Thanks. I reactivated the 144 kHz version. I am doing more data engineering to improve the classifier currently.

**Reply by rdz-oss** [reply]
I made a mistake when changing to 1444kHz version. In effect, it was running a hybrid between 144 and 256 kHz. Should be fixed now. My appologies. Now it should work .

**Comment by rdz-oss** [comment]
INFO on integration of BattyBirdNET-Analyzer to a PI

To everyone interested in getting a BirdNET-PI for bats: I am making progress on the integration of the BattyBirdNET into a test system.
The approach is:

    Use wildlife acoustics echo meter touch 2 as mic
    Install BattyBirdNET-Analyzer on the PI (working)
    Set it up as via its server.py as a systemd service (working)
    Install a modified version of BirdNET-Pi at 256kHz (working, buggy at this point)
    Analyze the audio stream for bats via a BirdNET-PI-server.py to BattyBirdNET-Analyzer-server.py bridge
    Status: they find each other, can call each other, but the interface is not as expected and needs tailoring (TODO)

Then, in principle, it should work. As I am working on this on the weekends while the kids jump around process is slow but steady.
If you want to check it out or help get the system up you can look/contribute here:

https://github.com/rdz-oss/BattyBirdNET-Pi

BR
Richard

**Comment by rdz-oss** [comment]
### News: The BattyBirdNET-Pi prototype is running end2end now!

It has been tricky but I am happy to say that the proof-of-concept stage has been reached.
However, as there are too  random bugs at this point, not yet fit for use of people who do not like to experiment and can handle some frustration. But I am confident we will get there. The ‘heavy-lifting’ has been achieved. Still much to do, but mostly bug hunting. 

What I have learned:

- Bat identification based on transfer learning on BirdNET-Analyzer is feasible - see BattyBirdNET-Analyzer
- BirdNet-Pi can be modified to run on 256kHz at 0.5625s - see BattyBirdNET-PI
- BattyBirdNET-PI can use BattyBirdNET-Analyzer to scan for bat calls

What works now on the prototype BattyBirdNET-Pi

- Use of Wildlife Acoustic Echo Meter Touch 2 directly to RASPI USB (via USB-USB cable)
- Overview page and all elements working (spectrogram to 125 kHz, recordings and current stats)
- Live sound (unaltered from mic)
- Todays Detections 
- Spectrogram
- Best recordings
- Species Stats
- Daily Charts
- Recordings
- View Log
- Tools (only tested partially, a lot of work needed e.g. on settings etc)

What is not good yet?

- Random bugs pop up - much work to do
- Cricket noise in my backyard classified as some Bats
- Tools - update, settings, bird weather integration ,…) much work to do
- Only implemented for Bavarian classifier as of today

Thank you all for your support  in getting to this point! Let’s go together, step by step, until it is production ready. Help, if you can, by providing bat call data and noise samples to BattyBirdNET-Analyzer, as well as feedback on issues and pull-requests.

If you are not a developer or frustration tolerant, wait a few more weeks.

Best regards,
Richard


Links
https://github.com/rdz-oss/BattyBirdNET-Analyzer
https://github.com/rdz-oss/BattyBirdNET-Pi

Prototype install (only if you can handle the pain of bugs AND live in the EU):

```  sh
curl -s https://raw.githubusercontent.com/rdz-oss/BattyBirdNET-Pi/main/newinstaller.sh | bash
``` 
<img width="1003" alt="Screenshot 2023-10-03 at 11 34 50" src="https://github.com/rdz-oss/BattyBirdNET-Analyzer/assets/63958211/11f14d4a-5850-47b6-a707-362498cfa65b">

<img width="913" alt="Screenshot 2023-10-03 at 11 35 06" src="https://github.com/rdz-oss/BattyBirdNET-Analyzer/assets/63958211/9b171e48-4c44-4d74-b16e-0126359fe28b">

<img width="1333" alt="Screenshot 2023-10-03 at 11 34 21" src="https://github.com/rdz-oss/BattyBirdNET-Analyzer/assets/63958211/61d0df8c-4a0f-4204-af8e-18601f58bfa2">

<img width="1319" alt="Screenshot 2023-10-03 at 11 35 27" src="https://github.com/rdz-oss/BattyBirdNET-Analyzer/assets/63958211/75a24cc7-7935-4669-a70b-70ef2238cb29">


<img width="1000" alt="Screenshot 2023-10-03 at 11 35 52" src="https://github.com/rdz-oss/BattyBirdNET-Analyzer/assets/63958211/76491014-4015-416f-8123-c277a60c2db7">

**Reply by rdz-oss** [reply]
The first bat caught with the BattyBirdNET-Pi prototype:

<img width="790" alt="Screenshot 2023-10-04 at 19 43 20" src="https://github.com/rdz-oss/BattyBirdNET-Analyzer/assets/63958211/317f7dc0-0a32-49ea-a57f-d701a098ea8d">

**Comment by rdz-oss** [comment]
INFO: I updated the training set for the BAVARIA classifier to better detect pipistrellus pipistrellus. It was shadowed by nathusis pipistrelle. Is much better now in keeping those apart.
Also, I added mor myotis myotis calls to reduce confusion with myotis dasycmene.
FYI, let me know if you run into issues,
Richard
